src/main/scala/com/fulcrumgenomics/alignment/NeedlemanWunschAligner.scala

```diff
@@ -116,66 +136,132 @@ class NeedlemanWunschAligner(val scoringFunction: (Byte,Byte) => Int,
+          matrices(Up).trace(i, 0)   = if (i == 1) Diagonal else Up
         }
       case Local =>
         forloop(from=1, until=query.length+1) { i =>
-          trace(i, 0)   = Done
-          scoring(i, 0) = 0
+          forloop(from=1, until=query.length+1) { i =>
+            LeftAndUp.foreach { direction =>
+              matrices(direction).scoring(i, 0) = MinStartScore
+              matrices(direction).trace(i, 0)   = Done
+            }
+            matrices(Diagonal).scoring(i, 0) = 0
+            matrices(Diagonal).trace(i, 0)   = Done
+          }
         }
     }
 
     // Along the top - for global we have to keep going left, for glocal/local we're done
     mode match {
       case Global =>
         forloop(from=1, until=target.length+1) { j =>
-          trace(0, j)   = Left
-          scoring(0, j) = scoring(0, j-1) + gapExtend + (if (j == 1) gapOpen else 0)
+          UpAndDiagonal.foreach { direction =>
+            matrices(direction).scoring(0, j) = MinStartScore
+            matrices(direction).trace(0 , j)   = Done
+          }
+          matrices(Left).scoring(0, j) = matrices(Left).scoring(0, j-1) + gapExtend + (if (j == 1) gapOpen else 0)
+          matrices(Left).trace(0, j)   = if (j == 1) Diagonal else Left
         }
       case Glocal | Local =>
         forloop(from=1, until=target.length+1) { j =>
-          trace(0, j)   = Done
-          scoring(0, j) = 0
+          LeftAndUp.foreach { direction =>
+            matrices(direction).scoring(0, j) = MinStartScore
+            matrices(direction).trace(0, j)   = Done
+          }
+          matrices(Diagonal).scoring(0, j) = 0
+          matrices(Diagonal).trace(0, j)   = Done
         }
     }
 
     // The interior of the matrix
-    forloop(from=1, until=query.length+1) { i =>
-      forloop(from=1, until=target.length+1) { j =>
-        val diagonalScore = scoring(i-1, j-1) + scoringFunction(query(i-1), target(j-1))
-        val upScore       = scoring(i-1, j  ) + gapExtend + (if (trace(i-1, j  ) == Up) 0 else gapOpen)
-        val leftScore     = scoring(i,   j-1) + gapExtend + (if (trace(i,   j-1) == Left) 0 else gapOpen)
-
-        val maxScore = max(max(diagonalScore, upScore), leftScore)
-        scoring(i,j) = maxScore
-
-        // Prefer mismatches first when back-tracing so as to push indels to the left of the alignment
-        // The choice to prefer deletions before insertions is largely arbitrary!
-        trace(i,j)   = if (maxScore == diagonalScore) Diagonal else if (maxScore == leftScore) Left else Up
-        // For local mode, we can start anywhere
-        if (mode == Local && scoring(i,j) < 0) {
-          scoring(i,j) = 0
-          trace(i,j)   = Done
+    var i = 1
+    while (i <= query.length) {
+      var j = 1
+      while (j <= target.length) {
+        { // Diagonal matrix can come from the previous diagonal, up, or left
+          val addend = scoringFunction(query(i-1), target(j-1))
```

Comment:
Could cache/store `query(i-1)` in a local variable outside the inner loop.

Reply:
Done.